<a href="https://colab.research.google.com/github/AmanShaikh7/Document-Retrival/blob/main/retriving_wikipedia_articles.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
!pip install turicreate

     |████████████████████████████████| 92.0 MB 5.3 kB/s 
     |████████████████████████████████| 3.5 MB 50.9 MB/s 
     |████████████████████████████████| 3.6 MB 55.6 MB/s 
     |████████████████████████████████| 322 kB 66.7 MB/s 
     |████████████████████████████████| 86.4 MB 58 kB/s 
     |████████████████████████████████| 18.3 MB 78 kB/s 
     |████████████████████████████████| 20.1 MB 58.0 MB/s 
     |████████████████████████████████| 3.8 MB 47.5 MB/s 
     |████████████████████████████████| 2.9 MB 53.0 MB/s 
     |████████████████████████████████| 50 kB 9.4 MB/s 
     |████████████████████████████████| 449 kB 66.5 MB/s 
  Created wheel for prettytable: filename=prettytable-0.7.2-py3-none-any.whl size=13714 sha256=6b3db1f1e85e50ed99a465d59f7f3422beecd71c3e385353ddcd76ed67bcb24f
  Stored in directory: /root/.cache/pip/wheels/b2/7f/f6/f180315b584f00445045ff1699b550fa895d09471337ce21c6
  Created wheel for resampy: filename=resampy-0.2.1-py3-none-any.whl size=320860 sha256=f4691c7f24

In [6]:
from google.colab import drive
import turicreate as tc

In [7]:
drive.mount('/content/drive')

Mounted at /content/drive


In [8]:
people='/content/drive/MyDrive/datasets/people_wiki.csv'

In [11]:
people = tc.SFrame('people_wiki.csv')

Finished parsing file /content/people_wiki.csv

Parsing completed. Parsed 100 lines in 0.990436 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 26690 lines. Lines per second: 27821.4

Finished parsing file /content/people_wiki.csv

Parsing completed. Parsed 38216 lines in 1.09978 secs.

In [12]:
people

Columns:
	URI	str
	name	str
	text	str

Rows: 38216

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[38216 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [14]:
john =people[people['name']=='Elton John']

In [15]:
john

Columns:
	URI	str
	name	str
	text	str

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [16]:
john['word_count']=tc.text_analytics.count_words(john['text'])

In [17]:
john

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 1

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'movements': 1.0, 'social... |
+-------------------------------+
[1 rows x 4 columns]

In [20]:
john_word_count_table=john[['word_count']].stack('word_count',new_column_name=['word','count'])

In [22]:
john_word_count_table

Columns:
	word	str
	count	float

Rows: 255

Data:
+-----------+-------+
|    word   | count |
+-----------+-------+
| movements |  1.0  |
|   social  |  1.0  |
|  champion |  1.0  |
|    wed    |  1.0  |
|   legal   |  1.0  |
|   became  |  1.0  |
|   after   |  1.0  |
|    2005   |  1.0  |
|  december |  2.0  |
|  furnish  |  2.0  |
+-----------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [23]:
john_word_count_table.sort('count',ascending=False)

Columns:
	word	str
	count	float

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |  27.0 |
|   in  |  18.0 |
|  and  |  15.0 |
|   of  |  13.0 |
|   a   |  10.0 |
|  has  |  9.0  |
|   he  |  7.0  |
|  john |  7.0  |
|   on  |  6.0  |
| since |  5.0  |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

now lets calculate tf idf

In [25]:
people['word_count'] = tc.text_analytics.count_words(people['text'])
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'melbourne': 1.0, 'parade... |
| alfred j lewy aka sandy le... | {'time': 1.0, 'each': 1.0,... |
| harpdog brown is a singer ... | {'society': 1.0, 'hamilton... |
| franz rottensteiner born i... | {'kurdlawitzpreis': 1.0, '... |
| henry krvits born 30 decem... | {'curtis': 1.0, '2007': 1.... |
| sam henderson born october... | {'asses': 1.0, 'sic': 1.0,... |
| aaron lacrate is an americ... | {'streamz': 1.0, 'includin... |
| trevor ferguson aka john f... | {'concordia': 1.0, 'creati... |
| grant nelson born 27 april... | {'heavies': 1.0, 'new': 1.... |
| cathy caruth born 1955 is ... | {'2002': 1.0, 'harvard': 1... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

In [26]:
tfidf = tc.text_analytics.tf_idf(people['word_count'])
tfidf

dtype: dict
Rows: 38216
[{'melbourne': 3.9072198219221255, 'parade': 5.488414522042831, 'education': 2.455410853691608, 'physical': 4.020131927343913, 'teaches': 3.7430746113698716, 'currently': 1.6286179577739404, 'also': 0.4610839302278755, 'coburg': 7.778420832830016, 'has': 0.42870780938068925, '2014': 2.2069799826627494, 'since': 1.4429234171323715, 'premiership': 4.552072993123115, 'division': 2.7973860085100517, '2011': 1.7002053593133795, 'edflhe': 10.551009555069799, 'district': 2.77657404476684, '2013': 1.9575963377421512, 'until': 1.7504435627898778, '2009': 1.5699565691683142, 'from': 0.5887365014438262, 'retiring': 3.8022500075781185, 'hill': 3.8128570604738408, 'coached': 4.017220717136454, 'box': 4.574658645771864, 'strathmore': 8.248424462075752, '2008': 1.5137139031073605, '2007': 1.4871989757895123, 'assistant': 2.5433095421857725, 'hawks': 5.653169755118887, '2006': 1.517883802519425, 'vflaffiliate': 7.3729557247218525, 'acted': 4.124521097612107, 'carltons': 7.02464

In [27]:
people['tfidf'] = tfidf

In [28]:
people

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 38216

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'melbourne': 1.0, 'parade... |
| alfred j lewy aka sandy le... | {'time': 1.0, 'each': 1.0,... |
| harpdog brown is a singer ... | {'society': 1.0, 'hamilton... |
| franz rottensteiner born i... | {'kurdlawitzpreis': 1.0, '... |
| henry krvits born 30 decem... | {'curtis': 1.0, '2007': 1.... |
| sam henderson born october... | {'asses': 1.0, 'sic': 1.0,... |
| aaron lacrate is an americ... | {'streamz': 1.0, 'includin... |
| trevor ferguson aka john f... | {'concordia': 1.0, 'creati... |
| grant nelson born 27 april... | {'heavies': 1.0, 'new': 1.... |
| cathy caruth born 1955 is ... | {'2002': 1.0, 'harvard': 1... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'melbourne': 3.9072198219... |
| {'time': 1.334190216995906... |
| {'society': 2.444796652449... |
| {'kurdlawitzpreis': 10.551... |
| {'curtis': 5.3092625400101... |
| {'asses': 9.45239726640168... |
| {'streamz': 10.55100955506... |
| {'concordia': 6.2743434360... |
| {'heavies': 8.759250085841... |
| {'2002': 1.878523479247525... |
+-------------------------------+
[38216 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [31]:
john = people[people['name']=='Elton John']
john

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'movements': 1.0, 'social... | {'movements': 5.0295486372... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [32]:
john[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 255

Data:
+---------------+--------------------+
|      word     |       tfidf        |
+---------------+--------------------+
|    furnish    | 18.904794532803376 |
|     elton     | 17.52443806127239  |
|   billboard   | 17.297891543129328 |
|      john     | 13.975644052369045 |
|  songwriters  | 11.176329849619782 |
| tonightcandle | 10.551009555069799 |
|  overallelton | 10.551009555069799 |
|   fivedecade  | 10.551009555069799 |
|      aids     | 10.296664346395037 |
|    19702000   | 9.857862374509853  |
+---------------+--------------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

**results 1**

**word count  **  
1) the	27.0
2) in	18.0
3) and	15.0   

 **tfidf**
1) furnish	18.904794532803376
2) elton	17.52443806127239
3) billboard	17.297891543129328

In [ ]:
bechamen =people[people['name']=='Victoria Beckham']
bechamen

In [35]:
paul =people[people['name']=='Paul McCartney']

now lets train our model using Knn


In [42]:
knn_model1 = tc.nearest_neighbors.create(people,features=['tfidf'],label='name',distance='cosine')

Starting brute force nearest neighbors model training.

Validating distance components.

Initializing model data.

Initializing distances.

Done.

In [43]:
knn_model2 =  tc.nearest_neighbors.create(people,features=['word_count'],label='name',distance='cosine')

Starting brute force nearest neighbors model training.

Validating distance components.

Initializing model data.

Initializing distances.

Done.

In [44]:
knn_model2.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00261671  | 6.166ms      |

| Done         |         | 100         | 258.733ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+-----------------------+------+
| query_label | reference_label |        distance       | rank |
+-------------+-----------------+-----------------------+------+
|      0      |    Elton John   | 2.220446049250313e-16 |  1   |
|      0      |  Sandro Petrone |  0.16822542751041114  |  2   |
|      0      |   Rod Stewart   |  0.16832716558706107  |  3   |
|      0      |  Roger Daltrey  |  0.17755418466559603  |  4   |
|      0      |   Peter Paret   |   0.1807348374034532  |  5   |
+-------------+-----------------+-----------------------+------+
[5 rows x 4 columns]

In [45]:
knn_model1.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00261671  | 7.257ms      |

| Done         |         | 100         | 270.581ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+------------------------+------+
| query_label | reference_label  |        distance        | rank |
+-------------+------------------+------------------------+------+
|      0      |    Elton John    | -2.220446049250313e-16 |  1   |
|      0      |   Rod Stewart    |   0.7161005679904313   |  2   |
|      0      |  George Michael  |   0.7456787524889006   |  3   |
|      0      | Sting (musician) |   0.7472178292480971   |  4   |
|      0      |   Phil Collins   |   0.7501078948582134   |  5   |
+-------------+------------------+------------------------+------+
[5 rows x 4 columns]